### ISSUE TITLE
Issues when using sourcegraph with a screen reader

### ISSUE THREAD

**Issue by mikolysz** [sourcegraph/sourcegraph#905]
- **Sourcegraph version:** sourcegraph.com
- **Platform information:** win 7 64 bit, chrome latest stable, NVDA 2018.3.2

#### Steps to reproduce:

1. Access sourcegraph.com with a screen reader. A screen reader is a piece of software that reads the content on the screen for users who are unable to see it. Scren readers are mostly used by blind and low vision people.
2. Navigate around the page, try performing various activities.

#### Expected behavior:
All actions a sighted person is able to do with a mouse should be doable with a screen reader.

#### Actual behavior:
- There are a few unlabelled buttons on the page. An unlabelled button is one that doesn't have a text description. Usually those buttons have a graphical icon instead of text. To fix this, all buttons that contain icons should be rewritten so that the icon is hidden for screenreaders with the aria-hidden attribute, and there's a span inside that button with a .sronly class (visible only for screen reader users) with a descriptive text of what that button does.
- There should be a way to access the hover tooltips without actually using a mouse and hovering it over some piece of text. I don't know how to implement this, I think a discussion is in order, maybe someone who knows more about accessibility from a frontend web developer's point of view will have an idea. I'm thinking along the lines of "detect where the focus is and allow invoking the tooltip with a shortcut" but I don't know how reliable that would be.
- It's impossible to click a symbol with a screen reader. Pressing enter on a symbol does nothing. This should somehow be fixed (though the symbols should just be made clickable, not turned into links). Turning them to links or giving them a special role that would make screen readers think they're links wouldn't be a good idea, as screen readers would say link before the name of each symbol (AKA extremely annoying). I don't know how to make sr users know that a symbol is clickable without saying clickable before the name of each symbol, which would be extremely annoying for advanced users.
- The code view is a table, where one column contains line numbers and the other contains the actual lines. A screen reader user expects that pressing down arrows will move them to the next line and up arrow will move to the pervious line. Nothing but the line contents should be announced when doing that, there should be a way to get the line number upon request. The current hehavior is that pressing down arrow moves you to a line number, the to the contents of that line, then to the number of the next line, then to its contents etc. Some screenreaders also announce table coordinates (i.e. row 14, column 2 func main() {) instead of just func main () {.
- There are probably manu more issues, I haven't tested the website deeply enough to find all of them.

I'm willing to help out with this, test fixes and provide more info and context if there's interest.

**Comment by sqs** [comment]
Thank you for the detailed report! @francisschmaltz is working on this area and can triage and plan fixes here.

**Comment by francisschmaltz** [comment]
@devil418 Thanks for reporting these! 

I've created the following issue for Sourcegraph:
https://github.com/sourcegraph/sourcegraph/issues/972
https://github.com/sourcegraph/sourcegraph/issues/973
https://github.com/sourcegraph/sourcegraph/issues/974

If you want, feel free to submit PRs for code changes that you find to solve these, or other, issues.

**Comment by christinaforney** [comment]
Still important for a11y

**Comment by felixfbecker** [comment]
We've made a lot of progress this year on screen reader usability and towards WCAG compliance, so I'm closing this issue in favor of the WCAG tracking issue (which I think contains individual issues for what's listed in the issue here with a lot more discussion): https://github.com/sourcegraph/sourcegraph/issues/31476
We also have an issue template now to report specific accessibility issues 🙂
